File: sdk/python/tests/test_protocol.py
from __future__ import annotations

import sys
import unittest
from pathlib import Path

PACKAGE_ROOT = Path(__file__).resolve().parents[1]
try:
    import glyphastore  # noqa: F401
except ImportError:
    sys.path.insert(0, str(PACKAGE_ROOT / "src"))

from glyphastore.protocol import (  # noqa: E402
    REQUEST_HEADER_BYTES,
    RESPONSE_HEADER_BYTES,
    ROUTING_ALG_SIPHASH24_V1,
    Opcode,
    Status,
    WorkerRouting,
    decode_init_identity,
    decode_request,
    decode_response,
    encode_init_identity,
    encode_request,
    encode_response,
    hash_key_routing,
    siphash24,
    worker_for,
)


def fixture(name: str) -> bytes:
    path = PACKAGE_ROOT / "tests" / "fixtures" / name
    return bytes(int(token, 16) for token in path.read_text(encoding="utf-8").split())


def frames(corpus: bytes) -> list[bytes]:
    output: list[bytes] = []
    offset = 0
    while offset < len(corpus):
        size = int.from_bytes(corpus[offset : offset + 4], "little")
        output.append(corpus[offset : offset + size])
        offset += size
    return output


class ProtocolTests(unittest.TestCase):
    def test_request_encoder_matches_every_canonical_fixture(self) -> None:
        expected = frames(fixture("wire_requests_v2.hex"))
        encoded = [
            encode_request(Opcode.INIT, 1),
            encode_request(Opcode.PING, 2, value=b"\x00ping\xff"),
            encode_request(Opcode.GET, 3, key=b"get\x00key"),
            encode_request(
                Opcode.PUT,
                4,
                key=b"put\x00key",
                value=b"\x10\x20\xff",
                expire_at_ns=123_456_789,
            ),
            encode_request(Opcode.ERASE, 5, key=b"erase-key"),
            encode_request(Opcode.BIND_WORKER, 6, target_worker=2),
            encode_request(Opcode.HEALTH, 7),
            encode_request(Opcode.READY, 8),
            encode_request(Opcode.STATS, 9),
        ]
        self.assertEqual(encoded, expected)

    def test_request_decoder_round_trips_canonical_fixture(self) -> None:
        decoded = [decode_request(frame) for frame in frames(fixture("wire_requests_v2.hex"))]
        self.assertEqual([request.opcode for request in decoded], list(Opcode))
        self.assertEqual(decoded[1].value, b"\x00ping\xff")
        self.assertEqual(decoded[3].expire_at_ns, 123_456_789)
        self.assertEqual(decoded[5].target_worker, 2)
        reencoded = [
            encode_request(
                request.opcode,
                request.request_id,
                key=request.key,
                value=request.value,
                expire_at_ns=request.expire_at_ns,
                target_worker=request.target_worker,
            )
            for request in decoded
        ]
        self.assertEqual(reencoded, frames(fixture("wire_requests_v2.hex")))

    def test_response_decoder_matches_every_canonical_fixture(self) -> None:
        decoded = [decode_response(frame) for frame in frames(fixture("wire_responses_v2.hex"))]
        self.assertEqual([response.status for response in decoded], list(Status))
        self.assertEqual(decoded[0].value, b"GlyphaStore/2")
        self.assertEqual(decoded[6].owner_worker, 2)
        self.assertTrue(all(response.worker_count == 4 for response in decoded))

    def test_response_encoder_round_trips_canonical_fixture(self) -> None:
        decoded = [decode_response(frame) for frame in frames(fixture("wire_responses_v2.hex"))]
        reencoded = [
            encode_response(
                response.status,
                response.request_id,
                value=response.value,
                owner_worker=response.owner_worker,
                worker_count=response.worker_count,
                routing_epoch=response.routing_epoch,
            )
            for response in decoded
        ]
        self.assertEqual(reencoded, frames(fixture("wire_responses_v2.hex")))

    def test_decoder_rejects_noncanonical_and_truncated_frames(self) -> None:
        request = bytearray(frames(fixture("wire_requests_v2.hex"))[0])
        request[36] = 1
        with self.assertRaises(ValueError):
            decode_request(bytes(request))
        with self.assertRaises(ValueError):
            decode_request(bytes(request[: REQUEST_HEADER_BYTES - 1]))

        response = bytearray(frames(fixture("wire_responses_v2.hex"))[0])
        response[28] = 1
        with self.assertRaises(ValueError):
            decode_response(bytes(response))
        with self.assertRaises(ValueError):
            decode_response(bytes(response[: RESPONSE_HEADER_BYTES - 1]))

    def test_routing_is_binary_and_deterministic(self) -> None:
        self.assertEqual(worker_for(b"", 4), 1)
        self.assertEqual(worker_for(b"key\x00\xff", 4), worker_for(b"key\x00\xff", 4))

    def test_siphash24_matches_aumasson_bernstein_vectors(self) -> None:
        k0 = 0x0706050403020100
        k1 = 0x0F0E0D0C0B0A0908
        self.assertEqual(siphash24(b"", k0, k1), 0x726F_DB47_DD0E_0E31)
        self.assertEqual(siphash24(bytes([0, 1, 2]), k0, k1), 0x8567_6696_D7FB_7E2D)

    def test_keyed_routing_and_init_identity(self) -> None:
        keyed = WorkerRouting(algorithm=ROUTING_ALG_SIPHASH24_V1, seed=0x1111_2222_3333_4444)
        digest = hash_key_routing(b"tenant-a/orders/1", keyed)
        self.assertEqual(digest, 0x712F_CEC5_7AC8_4546)
        self.assertEqual(worker_for(b"tenant-a/orders/1", 8, keyed), 6)
        self.assertNotEqual(digest, hash_key_routing(b"tenant-a/orders/1"))

        plain = encode_init_identity()
        self.assertEqual(plain, b"GlyphaStore/2")
        self.assertEqual(decode_init_identity(plain), WorkerRouting())

        extended = encode_init_identity(
            WorkerRouting(algorithm=ROUTING_ALG_SIPHASH24_V1, seed=0xABCD_EF01_2345_6789)
        )
        self.assertEqual(len(extended), 26)
        self.assertEqual(
            decode_init_identity(extended),
            WorkerRouting(algorithm=ROUTING_ALG_SIPHASH24_V1, seed=0xABCD_EF01_2345_6789),
        )
        with self.assertRaises(ValueError):
            decode_init_identity(b"GlyphaStore/2\x00bad")
        with self.assertRaises(ValueError):
            decode_init_identity(b"GlyphaStore/3")

    def test_u64_fields_reject_out_of_range_values(self) -> None:
        with self.assertRaises(ValueError):
            encode_request(Opcode.PUT, -1, key=b"k", value=b"v")
        with self.assertRaises(ValueError):
            encode_request(Opcode.PUT, 1, key=b"k", value=b"v", expire_at_ns=-1)
        with self.assertRaises(ValueError):
            encode_response(Status.OK, 1, routing_epoch=-1)


if __name__ == "__main__":
    unittest.main()

File: sdk/ruby/test/test_protocol.rb
# frozen_string_literal: true

require "minitest/autorun"
require "pathname"
require "glypha_store"

class ProtocolTest < Minitest::Test
  FIXTURES = Pathname.new(__dir__).join("fixtures")

  def fixture(name)
    FIXTURES.join(name).read.split.map { |token| token.to_i(16) }.pack("C*")
  end

  def frames(corpus)
    output = []
    offset = 0
    while offset < corpus.bytesize
      size = corpus.byteslice(offset, 4).unpack1("L<")
      output << corpus.byteslice(offset, size)
      offset += size
    end
    output
  end

  def test_request_encoder_matches_fixtures
    expected = frames(fixture("wire_requests_v2.hex"))
    encoded = [
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::INIT, 1),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::PING, 2, value: "\x00ping\xff".b),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::GET, 3, key: "get\x00key".b),
      GlyphaStore::Protocol.encode_request(
        GlyphaStore::Protocol::Opcode::PUT, 4,
        key: "put\x00key".b, value: "\x10\x20\xff".b, expire_at_ns: 123_456_789
      ),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::ERASE, 5, key: "erase-key".b),
      GlyphaStore::Protocol.encode_request(
        GlyphaStore::Protocol::Opcode::BIND_WORKER, 6, target_worker: 2
      ),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::HEALTH, 7),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::READY, 8),
      GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::STATS, 9)
    ]
    assert_equal expected, encoded
  end

  def test_request_decoder_round_trips
    expected = frames(fixture("wire_requests_v2.hex"))
    reencoded = expected.map do |frame|
      decoded = GlyphaStore::Protocol.decode_request(frame)
      GlyphaStore::Protocol.encode_request(
        decoded.opcode, decoded.request_id,
        key: decoded.key, value: decoded.value,
        expire_at_ns: decoded.expire_at_ns, target_worker: decoded.target_worker
      )
    end
    assert_equal expected, reencoded
  end

  def test_response_round_trips
    expected = frames(fixture("wire_responses_v2.hex"))
    reencoded = expected.map do |frame|
      decoded = GlyphaStore::Protocol.decode_response(frame)
      GlyphaStore::Protocol.encode_response(
        decoded.status, decoded.request_id,
        value: decoded.value, owner_worker: decoded.owner_worker,
        worker_count: decoded.worker_count, routing_epoch: decoded.routing_epoch
      )
    end
    assert_equal expected, reencoded
  end

  def test_worker_routing_is_deterministic
    key = "session\x0042".b
    assert_equal GlyphaStore::Protocol.fnv1a64(key) % 4, GlyphaStore::Protocol.worker_for(key, 4)
    assert_equal GlyphaStore::Protocol.worker_for(key, 4), GlyphaStore::Protocol.worker_for(key, 4)
  end

  def test_rejects_noncanonical_reserved
    frame = GlyphaStore::Protocol.encode_request(GlyphaStore::Protocol::Opcode::PING, 1, value: "x".b)
    mutated = frame.dup
    mutated.setbyte(36, 1)
    assert_raises(ArgumentError) { GlyphaStore::Protocol.decode_request(mutated) }
  end

  def test_siphash24_paper_vectors
    k0 = 0x0706050403020100
    k1 = 0x0f0e0d0c0b0a0908
    assert_equal 0x726fdb47dd0e0e31, GlyphaStore::Protocol.siphash24("".b, k0, k1)
    assert_equal 0x85676696d7fb7e2d, GlyphaStore::Protocol.siphash24("\x00\x01\x02".b, k0, k1)
  end

  def test_keyed_routing_and_init_identity
    keyed = GlyphaStore::Protocol::WorkerRouting.new(algorithm: GlyphaStore::Protocol::ROUTING_ALG_SIPHASH24_V1, seed: 0x1111222233334444)
    digest = GlyphaStore::Protocol.hash_key_routing("tenant-a/orders/1".b, keyed)
    assert_equal 0x712fcec57ac84546, digest
    assert_equal 6, GlyphaStore::Protocol.worker_for("tenant-a/orders/1".b, 8, keyed)
    plain = GlyphaStore::Protocol.encode_init_identity
    assert_equal GlyphaStore::Protocol::IDENTITY.b, plain
    assert_equal false, GlyphaStore::Protocol.decode_init_identity(plain).keyed?
    extended = GlyphaStore::Protocol.encode_init_identity(
      GlyphaStore::Protocol::WorkerRouting.new(algorithm: GlyphaStore::Protocol::ROUTING_ALG_SIPHASH24_V1, seed: 0xABCDEF0123456789)
    )
    assert_equal 26, extended.bytesize
    decoded = GlyphaStore::Protocol.decode_init_identity(extended)
    assert_equal true, decoded.keyed?
    assert_equal 0xABCDEF0123456789, decoded.seed
    assert_raises(ArgumentError) { GlyphaStore::Protocol.decode_init_identity("GlyphaStore/2\x00bad".b) }
  end
end
